### DISCUSSION TITLE
How do I bind a Model (or a Collection) to a Chart?

### DISCUSSION THREAD

**Discussion by ultrablue** [maartenpaauw/laravel-charts-css#33] [Q&A] [answered]
Hi:
I see in the readme how to create a Chart via Artisan, and how to use it in a Blade template.
But the examples all seem to have hard-coded data in the datasets() method. 
How would I bind a Model (or a Collection) to have it's data represented as a Chart?

Thank you!

**Comment by maartenpaauw** [comment, marked as answer]
Good question! The library is not tightly coupled to an eloquent model. You are free how you return your datasets. Within the example it's hard coded, but you can also query the database within the datasets method and map it into dataset instances. Let me know if you need a bit of example code?

**Reply by ultrablue** [reply]
Thanks for the reply.
What makes sense to me is something like this:
The Controller invokes a Model and runs a query, resulting in a Collection.
Still in the Controller, I create a Chart and pass it the Collection I got in the previous step to Chart::datasets().

Am I on the right path?

It seems like that that use case will be very common for your Package, so yes, some example code will be helpful.

Thanks again!

**Reply by maartenpaauw** [reply]
Okay, if I've some time tomorrow I'll provide you some example code.

**Reply by ultrablue** [reply]
Thank you!

**Reply by maartenpaauw** [reply]
As promised!

Within your controller:

```php
$countries = Country::with('medals')->get();

return view('welcome', [
    'countries' => $countries,
]);
```

Within you view

```html
<x-medals-chart :countries="$countries" />
```

Within the MedalsChart component add a constructor so we can use the given countries property.

```php
/**
 * @var Collection
 */
private $countries;

public function __construct(Collection $countries)
{
    $this->countries = $countries;
}
```

and replace the datasets() method with:

```php
protected function datasets(): DatasetsContract
{
    return new Datasets(
        new Axes('Country', ['Gold', 'Silver', 'Bronze']),
        ...$this->countries->map(function (Country $country) {
            return new Dataset([
                new Entry(new Value($country->medals()->where('type', 'gold')->count())),
                new Entry(new Value($country->medals()->where('type', 'silver')->count())),
                new Entry(new Value($country->medals()->where('type', 'bronze')->count())),
            ], new Label($country->name));
        })->toArray(),
    );
}
```

It's not the most optimised code but its for illustration only.

I hope it's clear how to create charts for eloquent models.

**Reply by ultrablue** [reply]
Thank you! I suspect that this will come in handy for many developers.

**Reply by maartenpaauw** [reply]
Thanks! I think I need to add this to the README. I'm to convert the whole README file to a wiki because its getting large. When that's done i'll add Eloquent examples.
